File: apps/michi/src/lib/api/rituals.ts
/**
 * Rituals API Client
 * Shinkofa Platform - Frontend
 */

import apiClient from './client'
import type { Ritual, RitualFilters, CreateRitualInput, UpdateRitualInput } from '@/types/api'

/**
 * Get all rituals with optional filters
 */
export async function getRituals(filters?: RitualFilters): Promise<Ritual[]> {
  const params = new URLSearchParams()
  if (filters?.category) params.append('category', filters.category)
  if (filters?.completed_today !== undefined)
    params.append('completed_today', filters.completed_today.toString())

  const url = `/rituals${params.toString() ? `?${params.toString()}` : ''}`
  const response = await apiClient.get<Ritual[]>(url)
  return response.data
}

/**
 * Get single ritual by ID
 */
export async function getRitual(ritualId: string): Promise<Ritual> {
  const response = await apiClient.get<Ritual>(`/rituals/${ritualId}`)
  return response.data
}

/**
 * Create new ritual
 */
export async function createRitual(input: CreateRitualInput): Promise<Ritual> {
  const response = await apiClient.post<Ritual>('/rituals', input)
  return response.data
}

/**
 * Update existing ritual
 */
export async function updateRitual(
  ritualId: string,
  input: UpdateRitualInput
): Promise<Ritual> {
  const response = await apiClient.put<Ritual>(`/rituals/${ritualId}`, input)
  return response.data
}

/**
 * Delete ritual
 */
export async function deleteRitual(ritualId: string): Promise<void> {
  await apiClient.delete(`/rituals/${ritualId}`)
}

/**
 * Reset all rituals (set completed_today to false)
 */
export async function resetRituals(): Promise<{ message: string; reset_count: number }> {
  const response = await apiClient.post<{ message: string; reset_count: number }>('/rituals/reset')
  return response.data
}

File: apps/api-shizen/app/routes/rituals.py
"""
Rituals CRUD endpoints
Shinkofa Platform - Planner
"""
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
import uuid

from app.core.database import get_db
from app.models.ritual import Ritual
from app.schemas.ritual import Ritual as RitualSchema, RitualCreate, RitualUpdate
from app.utils.auth import get_current_user_id

router = APIRouter(prefix="/rituals", tags=["rituals"])


@router.get("", response_model=List[RitualSchema])
def get_rituals(
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db),
    category: str = None,
):
    """
    Get all rituals for current user

    Query params:
    - category: Filter by category (morning, evening, daily, custom)
    """
    query = db.query(Ritual).filter(Ritual.user_id == user_id)

    if category:
        query = query.filter(Ritual.category == category)

    return query.order_by(Ritual.order, Ritual.label).all()


@router.get("/{ritual_id}", response_model=RitualSchema)
def get_ritual(
    ritual_id: str,
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    """Get a single ritual by ID"""
    ritual = (
        db.query(Ritual)
        .filter(Ritual.id == ritual_id, Ritual.user_id == user_id)
        .first()
    )

    if not ritual:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Ritual {ritual_id} not found",
        )

    return ritual


@router.post("", response_model=RitualSchema, status_code=status.HTTP_201_CREATED)
def create_ritual(
    ritual_data: RitualCreate,
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    """Create a new ritual"""
    new_ritual = Ritual(
        id=f"ritual-{uuid.uuid4()}",
        user_id=user_id,
        **ritual_data.model_dump(),
    )

    db.add(new_ritual)
    db.commit()
    db.refresh(new_ritual)

    return new_ritual


@router.put("/{ritual_id}", response_model=RitualSchema)
def update_ritual(
    ritual_id: str,
    ritual_update: RitualUpdate,
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    """Update an existing ritual"""
    ritual = (
        db.query(Ritual)
        .filter(Ritual.id == ritual_id, Ritual.user_id == user_id)
        .first()
    )

    if not ritual:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Ritual {ritual_id} not found",
        )

    # Update only provided fields
    update_data = ritual_update.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(ritual, field, value)

    db.commit()
    db.refresh(ritual)

    return ritual


@router.delete("/{ritual_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_ritual(
    ritual_id: str,
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    """Delete a ritual"""
    ritual = (
        db.query(Ritual)
        .filter(Ritual.id == ritual_id, Ritual.user_id == user_id)
        .first()
    )

    if not ritual:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Ritual {ritual_id} not found",
        )

    db.delete(ritual)
    db.commit()

    return None


@router.post("/reset", status_code=status.HTTP_200_OK)
def reset_rituals(
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    """Reset all rituals (mark as not completed)"""
    db.query(Ritual).filter(Ritual.user_id == user_id).update({"completed_today": False})
    db.commit()

    return {"message": "All rituals reset successfully"}
